src/local/room_position.rs

```diff
@@ -0,0 +1,375 @@
+/// `i32` in "non-human readable" formats like [`bincode`].
+///
+/// You can also pass `LocalRoomPosition` into JavaScript using the `js!{}`
+/// macro provided by `stdweb`, or helper methods using the same code like
+/// [`MemoryReference::set`][crate::memory::MemoryReference::set]. It will be
+/// serialized the same as in JSON, as an object with `roomName`, `x` and `y`
+/// properties.
+///
+/// *Note:* serializing using `js!{}` or `MemoryReference::set` will _not_
+/// create a `LocalRoomPosition`, only something with the same properties.
+///
+/// If you need a reference to a `RoomPosition` in JavaScript to use manually,
+/// you have two options:
+///
+/// - Use `.remote()` to get a `RoomPosition`, and then use that reference in
+///   JavaScript
+///
+/// - Convert the room position to an integer with
+///   [`LocalRoomPosition::packed_repr`], send that to JS, and use the
+///   `pos_from_packed` JavaScript function provided by this library:
+///
+///   ```no_run
+///   use stdweb::js;
+///   use screeps::LocalRoomPosition;
+///
+///   let pos = LocalRoomPosition::new(20, 21, "E5N6".parse().unwrap());
+///   let result = js! {
+///       let pos = pos_from_packed(@{pos.packed_repr()});
+///       pos.roomName
+///   };
+///   ```
+///
+/// [`bincode`]: https://github.com/servo/bincode
+/// [`RoomPosition::local`]: crate::RoomPosition::local
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct LocalRoomPosition {
+    /// A bit-packed integer, containing, from highest-order to lowest:
+    ///
+    /// - 1 byte: (room_x) + 128
+    /// - 1 byte: (room_y) + 128
+    /// - 1 byte: x
+    /// - 1 byte: y
+    ///
+    /// For `Wxx` rooms, `room_x = -xx - 1`. For `Exx` rooms, `room_x = xx`.
+    ///
+    /// For `Nyy` rooms, `room_y = -yy - 1`. For `Syy` rooms, `room_y = yy`.
+    ///
+    /// This is the same representation used in the Screeps server, allowing for
+    /// easy translation. Besides the method names and display representation,
+    /// this is the one part of LocalRoomPosition copied directly from the
+    /// engine code.
+    packed: u32,
+}
+
+impl fmt::Debug for LocalRoomPosition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LocalRoomPosition")
+            .field("packed", &self.packed)
+            .field("x", &self.x())
+            .field("y", &self.y())
+            .field("room_name", &self.room_name())
+            .finish()
+    }
+}
+
+impl fmt::Display for LocalRoomPosition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[room {} pos {},{}]",
+            self.room_name(),
+            self.x(),
+            self.y()
+        )
+    }
+}
+
+impl LocalRoomPosition {
+    /// Create a new LocalRoomPosition
+    ///
+    /// # Panics
+    ///
+    /// Will panic if either `x` or `y` is larger than 49, or if `room_name` is
+    /// outside of the range `E127N127 - W127S127`.
```

Comment:
I believe the first pair is 128?

Reply:
The maximum room coordinate is 128, but the fact of there being multiple "0" rooms (`E0N0 != W0N0`) means the actual room name of both maximum and minimum is `127`.